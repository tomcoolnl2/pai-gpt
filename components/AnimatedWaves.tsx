import * as React from 'react';
import * as Three from 'three';
import * as Fiber from '@react-three/fiber';
import circleImg from 'public/circle.png';

export const AnimationPoints: React.FC = () => {
	//
	const bufferRef = React.useRef<Three.BufferAttribute | null>(null);
	const imgTexture = Fiber.useLoader(Three.TextureLoader, circleImg.src);

	let t = 0;
	const frequency = 0.002;
	const amplitude = 3;

	const graph = React.useCallback(
		(x: number, z: number) => Math.sin(frequency * (x ** 2 + z ** 2 + t)) * amplitude,
		[t, frequency, amplitude],
	);

	const count = 500;
	const sep = 3;

	const bufferAttrPositions = React.useMemo(() => {
		const positions = [];
		for (let xi = 0; xi < count; xi += 1) {
			for (let zi = 0; zi < count; zi += 1) {
				let x = sep * (xi - count / 2);
				let z = sep * (zi - count / 2);
				let y = graph(x, z);
				positions.push(x, y, z);
			}
		}
		return new Float32Array(positions);
	}, [graph]);

	Fiber.useFrame(() => {
		const positions = bufferRef.current.array;
		t += 5;
		let i = 0;
		for (let xi = 0; xi < count; xi += 1) {
			for (let zi = 0; zi < count; zi += 1) {
				let x = sep * (xi - count / 2);
				let z = sep * (zi - count / 2);
				positions[i + 1] = graph(x, z);
				i += 3;
			}
		}
		bufferRef.current.needsUpdate = true;
	});

	return (
		<points>
			<bufferGeometry attach="geometry">
				<bufferAttribute
					ref={bufferRef}
					attach="attributes-position"
					array={bufferAttrPositions}
					count={bufferAttrPositions.length / 3}
					itemSize={3}
				/>
			</bufferGeometry>
			<pointsMaterial
				attach="material"
				map={imgTexture}
				color={0x00aaff}
				size={0.5}
				sizeAttenuation
				transparent={false}
				alphaTest={0.5}
				opacity={0.5}
			/>
		</points>
	);
};

export const AnimatedWaves: React.FC = () => {
	return (
		<React.Suspense fallback={null}>
			<Fiber.Canvas camera={{ position: [100, 10, 0], fov: 75 }}>
				<React.Suspense fallback={null}>
					<AnimationPoints />
				</React.Suspense>
			</Fiber.Canvas>
		</React.Suspense>
	);
};
