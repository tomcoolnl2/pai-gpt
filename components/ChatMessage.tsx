import React from 'react';
import Image from 'next/image';
import { useUser } from '@auth0/nextjs-auth0/client';
import { ConversationRole, Message } from 'context/conversation/ChatMessage';
import logo from 'public/logo.svg';

type Props = Omit<Message, 'id'>;

export const ChatMessage: React.FC<Props> = ({ role, content }) => {
	//
	const { user } = useUser();

	console.log(role);

	const avatar = React.useMemo(() => {
		const defaultSrc = 'http://www.gravatar.com/avatar';
		let src = '';
		switch (role) {
			case ConversationRole.ASSISTANT:
			case ConversationRole.SYSTEM:
				src = logo.src;
				break;
			case ConversationRole.USER:
				src = user ? user.picture : defaultSrc;
				break;
			default:
				src = defaultSrc;
		}
		return <Image src={src} width={30} height={30} alt={`${role} avatar`} className="avatar" />;
	}, [user, role]);

	return (
		<li className={`message message-${role}`}>
			{avatar}
			<p>{content}</p>
		</li>
	);
};
