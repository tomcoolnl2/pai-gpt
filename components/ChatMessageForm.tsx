import React from 'react';
import { AnswerMessage, SystemErrorMessage } from 'model';
import { useConversationContext } from 'context/conversation';

export const ChatMessageForm: React.FC<{ conversationId: null | string }> = ({ conversationId }) => {
	//
	const [question, setQuestion] = React.useState<string>('');
	const { answerStream, sendMessage } = useConversationContext();

	const disabled = React.useMemo(() => {
		if (answerStream instanceof SystemErrorMessage) {
			return true;
		}
		if (answerStream instanceof AnswerMessage) {
			return !answerStream.done;
		}
		return false;
	}, [answerStream]);

	const handleSubmit = React.useCallback(
		async (event: React.FormEvent<HTMLFormElement>) => {
			event.preventDefault();
			sendMessage(conversationId, question);
			setQuestion('');
		},
		[question, sendMessage],
	);

	const handleKeyPress = React.useCallback(
		(event: React.KeyboardEvent<HTMLTextAreaElement>) => {
			if (event.key === 'Enter' && !event.shiftKey) {
				handleSubmit(event as unknown as React.FormEvent<HTMLFormElement>);
			}
		},
		[handleSubmit],
	);

	return (
		<footer className="bg-gray-800 p-10">
			<form onSubmit={handleSubmit} noValidate>
				<fieldset className="flex gap-2" disabled={disabled}>
					<textarea
						value={question}
						onChange={(e) => setQuestion(e.target.value)}
						onKeyDown={handleKeyPress}
						placeholder="Send a message..."
						className="textarea"
					/>
					<button type="submit" className="btn">
						Send
					</button>
				</fieldset>
			</form>
		</footer>
	);
};
