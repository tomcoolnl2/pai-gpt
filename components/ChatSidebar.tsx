import Link from 'next/link';

export const ChatSidebar: React.FC = () => {
	return (
		<aside className="bg-gray-900 text-white p-10">
			<Link href="/api/auth/logout">Logout</Link>
		</aside>
	);
};
