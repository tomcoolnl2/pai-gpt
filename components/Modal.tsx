import React from 'react';
import ReactDOM from 'react-dom';
import { faClose } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface Props {
	open: boolean;
	dismissable?: boolean;
	handleClose: () => void;
	children: React.ReactNode;
}

export const Modal: React.FC<Props> = ({ open, dismissable = true, handleClose, children }) => {
	//
	const [documentMounted, setDocumentMounted] = React.useState<boolean>(false);

	React.useEffect(() => {
		setDocumentMounted(true);
	}, []);

	return (
		documentMounted &&
		ReactDOM.createPortal(
			<div onClick={handleClose} className={`modal-backdrop ${open ? 'open' : 'invisible'}`}>
				<div onClick={(e) => e.stopPropagation()} className={`modal${open ? ' open' : ''}`}>
					{dismissable && (
						<button onClick={handleClose} className="modal-close">
							<FontAwesomeIcon icon={faClose} />
						</button>
					)}
					{children}
				</div>
			</div>,
			document.getElementById('modal-root'),
		)
	);
};
