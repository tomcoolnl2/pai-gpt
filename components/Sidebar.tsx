import Link from 'next/link';
import React from 'react';
import { useConversationContext } from 'context/conversation';
import { Conversation } from 'model';

export const Sidebar: React.FC = () => {
	//
	const { conversations } = useConversationContext();

	return (
		<aside className="bg-gray-900 text-white">
			<Link className="btn" href="/chat">
				New Chat
			</Link>
			<ul className="flex-1 overflow-auto bg-gray-950">
				{conversations &&
					conversations.map((conversation) => (
						<li key={conversation.id} className="overflow-auto">
							<code>{conversation.id}</code>
							{conversation.title}
						</li>
					))}
			</ul>
			<Link href="/api/auth/logout">Logout</Link>
		</aside>
	);
};
