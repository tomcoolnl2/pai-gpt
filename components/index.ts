export * from './AnimatedWaves';
export * from './ChatSidebar';
export * from './ChatMessageForm';
export * from './ChatMessage';
