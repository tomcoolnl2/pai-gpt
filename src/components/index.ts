export * from './AnimatedWaves';
export * from './Spinner';
export * from './Sidebar';
export * from './SidebarItem';
export * from './Logo';
export * from './ChatMessageForm';
export * from './ChatMessage';
export * from './Modal';
export * from './DeleteConversationModal';
