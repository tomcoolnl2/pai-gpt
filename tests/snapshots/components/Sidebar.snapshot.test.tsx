import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import { Sidebar } from 'components';

describe('Sidebar', () => {
	beforeAll(() => {
		jest.spyOn(ReactDOM, 'createPortal').mockImplementation((element) => {
			return element as React.ReactPortal;
		});
	});

	afterEach(() => {
		jest.restoreAllMocks();
	});

	it('matches snapshot', () => {
		const { container } = render(<Sidebar conversationId="1" />);
		expect(container).toMatchSnapshot();
	});
});
