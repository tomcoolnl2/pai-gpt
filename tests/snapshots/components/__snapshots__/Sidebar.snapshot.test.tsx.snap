// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar matches snapshot 1`] = `
<div>
  <aside
    class="sidebar"
  >
    <a
      class="sidebar-item btn"
      href="/chat"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-plus "
        data-icon="plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
          fill="currentColor"
        />
      </svg>
      New Chat
    </a>
    <nav
      class="sidebar-list"
    >
      <ul>
        <li
          class="sidebar-item flex-col"
        >
          <div
            role="status"
          >
            <svg
              aria-hidden="true"
              class="spinner"
              fill="none"
              viewBox="0 0 100 101"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
                fill="currentColor"
              />
              <path
                d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
                fill="currentFill"
              />
            </svg>
            <span
              class="sr-only"
            >
              Loading...
            </span>
          </div>
        </li>
      </ul>
    </nav>
    <a
      class="sidebar-item"
      href="/api/auth/logout"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-from-bracket "
        data-icon="right-from-bracket"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M377.9 105.9L500.7 228.7c7.2 7.2 11.3 17.1 11.3 27.3s-4.1 20.1-11.3 27.3L377.9 406.1c-6.4 6.4-15 9.9-24 9.9c-18.7 0-33.9-15.2-33.9-33.9l0-62.1-128 0c-17.7 0-32-14.3-32-32l0-64c0-17.7 14.3-32 32-32l128 0 0-62.1c0-18.7 15.2-33.9 33.9-33.9c9 0 17.6 3.6 24 9.9zM160 96L96 96c-17.7 0-32 14.3-32 32l0 256c0 17.7 14.3 32 32 32l64 0c17.7 0 32 14.3 32 32s-14.3 32-32 32l-64 0c-53 0-96-43-96-96L0 128C0 75 43 32 96 32l64 0c17.7 0 32 14.3 32 32s-14.3 32-32 32z"
          fill="currentColor"
        />
      </svg>
      Logout
    </a>
  </aside>
  <div
    class="modal-backdrop invisible"
  >
    <div
      class="modal"
    >
      <button
        class="modal-close"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark "
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="text-center w-56"
      >
        <div
          class="mx-auto my-4 w48 text-white"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-trash "
            data-icon="trash"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
              fill="currentColor"
            />
          </svg>
          <h3
            class="text-lg"
          >
            Confirm Delete
          </h3>
          <p
            class="text-sm pb-6 pt-6"
          >
            Are you sure you want to delete this conversation?
          </p>
          <div
            class="flex gap-4"
          >
            <button
              class="btn btn-danger w-full"
            >
              Delete
            </button>
            <button
              class="btn btn-light w-full"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
